# Problem 2.C6 in Beer and Johnston

### Given:

A solid truncated cone is subjected to an axial force $P$ as shown below.

![problem diagram](problem_diagram.png)

When load $P$ is applied to the cone, the exact elongation, $\epsilon$ is:

$$ \epsilon = \frac{PL}{2\pi c^{2}E} $$

### Find:

By replacing the cone by $n$ circular cylinders of equal thickness, determine the elongation of the the truncated cone. Also calculate the percent error obtained from using $n=6$, $n=12$, and $n=60$ circular cylinders.

### Assume:

This problem can be completed generally, but for this solution, we are going to assume the following:

load $P=10 \ MPa$

minor radius $c=5 \ mm$

major radius $2c=10 \ mm$

length $L=100 \ mm$

Elastic Modulus $E=120 \ GPa$

### Solution:

We are going to use Python to code the solution to this problem. If you don't already have Python installed on your computer, I recommend installing the [Anaconda](https://www.anaconda.com/distribution/) distribution of Python. See [this post](https://pythonforundergradengineers.com/installing-anaconda-on-windows.html) to learn how to install Anaconda on your computer. 

I am coding this solution in a [Jupyter Notebook](https://jupyter.org/). Once you install Anaconda, you can open a Jupyter notebook from the Windows start menu or the Anaconda Prompt. See [this post](https://pythonforundergradengineers.com/opening-a-jupyter-notebook-on-windows.html) to learn about the ways to open a Jupyter notebook.

Alternatively, instead of using a Jupyter notebook, you could code your solution in a ```.py``` file.

**Alright. Let's get coding....**

Using the equation for the exact elongation, ($\epsilon$) above and the assumed values for $P$, $c$, $L$, and $E$, we can calculate $\epsilon$, which we will call ```e_exact```. The equation and assumed values are below.

$$ \epsilon = \frac{PL}{2\pi c^{2}E} $$

$P=10 \ MPa$

$c=5 \ mm$

$L=100 \ mm$

$E=120 \ GPa$

To code this solution in Python, first we need to import the ```math``` module. From the ```math``` module we will use ```math.pi``` as our value for $pi$. We can print our calculated ```e_exact``` using a Python _f-string_. Python f-strings allow us to put variables on the inside of print statements. Variables inside Python f-strings need to be surrounded by curly braces ```{ }```. Make sure to pre-pend the f-string with the letter ```"f"```.

In [2]:
import math

P = 10
c = 5
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

e_exact = (P*L)/(2*math.pi*(c**2)*E)

print(f"The exact value for delfection of the cone is d_exact = {e_exact}")

The exact value for delfection of the cone is d_exact = 5.305164769729845e-05


Next we'll see how the approximations of the elongation of the cone compare to the exact elongation ```e_exact``` calculated above. We'll calculate the approximate elongation by dividing the cone into separate "disks". Below is a picture of a children's toy to illustrate the concept of modeling a truncated cone as a set of separate disks.

![children's toy](child_toy.png)


For each "disk" the elongation $\delta$ is equal to:

$$ \delta = \frac{PL}{AE} $$

where

$\delta$ = deformation, or change in thickness of the disk

$P$ = load

$L$ = original length or thickness of the disk

$A$ = cross-sectional area of the disk

$E$ = Elastic Modulus (a material property)

We can plot the upper surface of our truncated cone in two dimensions. The x-axis on the plot is the cone length and the y-axis is the cone radius. The resulting plot looks like the hand-drawn plot below.

![r vs L](r_2r.png)

OK. Let's make our first approximation. If the truncated cone is modeled as a single disk (something that kind of looks like a can of soup), the radius of that single disk should be the average of the minor radius $r = 5 \ mm$ and the major radius $2r = 10 \ mm$. The plot below shows this approximation. Note how the upper surface of the cone (the line on the plot) has been approximated as a single horizontal line at a value of $r = 7.5 \ mm$.

![divide cone by 1](divide_by_n_equals1.png)

We can calculate the the elongation of this single cylinder with a radius of $r = 7.5$ using the standard equation for elongation in engineering mechanics.

$$ \delta_1 = \frac{PL}{A_{1}E} $$

$\delta_1$ = deformation, or change in thickness of the one disk

$P = 10 \ MPa$

$r_1 =7.5 \ mm$ (radius of the one disk)

$A_1 = \pi r^2$ (area of the one disk)

$L = 100 \ mm$

$E = 120 \ GPa$

Let's code this approximation of the elongation (modeling the truncated cone as a single thick disk) in Python.

In [4]:
import math

P = 10
r_1 = 7.5
A_1 = math.pi*r_1**2
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

d_1 = (P*L)/(A_1*E)

print(f"The approximate value for delfection of the cone modeled as one disk is d_1 = {d_1}")

The approximate value for delfection of the cone modeled as one disk is d_1 = 4.71570201753764e-05


The approximate value calculated above is around $4.72 \times 10^{-5}$. This approximate value is in the same order of magnitude as the previously calculated exact value of deflection ```d_exact``` which was about $5.31 \times 10^{-5}$.

We can more closely approximate the deformation by dividing our truncated cone into more numerous by successively thinner disks. The plot below shows the progression of using two disks, then three disks, then four disks.

![3 approximations](divide_by_n.png)

Note that as the number of disks increases the thickness of each disk also increases. The thickness of each disk $L$ can be calculated by taking the thickness (or Length) of the entire truncated cone and dividing it by the number of disks.

$$ L_n = \frac{L}{n} $$

$L_n$ = thickness of each disk

$L$ = thickness of the whole truncated cone

$n$ = number of disks the truncated cone is approximated by

Half of the disk thickness is where we will set our disk radius. If we take the $L$ value (x-value) as the right-hand edge of each disk, we need to back up half-way through the disk to set the location of the disk radius.

$$ L_i = \frac{n_i}{L} - (L_n/2) $$

$L_i$ = $L$ value where we set the radius for each disk

$n_i$ = disk number

$L$ = thickness of the whole truncated cone

$L_n$ = thickness of each disk

$L_n/2$ = half the thickness of each disk (need to back up half-way through a disk to arrive at the $L$ value to set the radius)

Great, we have an equation for the $L$ value of each disk, but we still need the radius of each disk to calculate the area of each disk and ultimately the deformation of each disk. 

What we need is an equation that spits out a radius $r$ given a length $L$. 

We can write the equation with the help of a plot. The plot below shows how radius $r$ changes based on length $L$. The relationship is a simple line that we can approximate using the form $y = mx+b$. Where $y = r$, $m = slope$, and $b = y-intercept$. The hand-drawn plot below shows the equation for our assumed values.

![plot with equation](r_2r_with_10_5_100.png)

So our equation for $r$ in terms of $L_i$ is:

$$ r = \frac{-5}{100}L_i + 10 $$

\frac{-5}{100} = slope of $r$ vs. $L$

$L_i$ = L value (distance "up" the truncated cone)

10 = y-intercept (the major radius)

Now let's use our formula for disk radius $r$ and approximate the elongation of the truncated cone using two disks ($n=2$).

In [27]:
import math

P = 10
c = 5
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

n = 2 # two disks
L_n = L/n #thickness of each disk

# calculate elgongation of the first disk (the disk with the larger radius)
i=1
L_1 = i*L/n-(L_n*(1/2))
print(f"L({i}) position: {L_1}")
r_1 = (-c/L)*L_1+2*c
print(f"r({i}) length: {r_1}")
A_1 = math.pi*r_1**2
d_1 = (P*L/n)/(A_1*E)
print(f"The delfection of the first disk d_1 = {d_1}")

# calculate elgongation of the second disk (the disk with the larger radius)
i=2
L_2 = i*L/n-(L_n*(1/2))
print(f"L({i}) position: {L_2}")
r_2 = (-c/L)*L_2+2*c
print(f"r({i}) length: {r_2}")
A_2 = math.pi*r_2**2
d_2 = (P*L/n)/(A_2*E)
print(f"The delfection of the second disk d_2 = {d_2}")

# add the two deflections together to calculate the total deflection
d_t = d_1 + d_2
print(f"The total estimated deflections using {n} disks is d_t = {d_t}")

L(1) position: 25.0
r(1) length: 8.75
The delfection of the first disk d_1 = 1.7322987003199492e-05
L(2) position: 75.0
r(2) length: 6.25
The delfection of the second disk d_2 = 3.395305452627101e-05
The total estimated deflections using 2 disks is d_t = 5.12760415294705e-05


Great! We can see our approximation using two disks which is about $5.13 \times 10^{-5}$ is closer to the exact value than our first approximation using one disk.

But hand coding in ```L_1```, ```r_1```, ```A_1```, ```d_1``` followed by ```L_2```, ```r_2```, ```A_2```, ```d_2``` even though we are doing the same operations each time is repetative. If we have 5 or 60 disks, we would end up with a lot of code. 

So to make our calculation work for $n$ disks, we'll use a _for loop_. We want to start our loop at ```i=1``` and end our loop at ```i=n```. For example, the code below shows how the calculation above can be refactored into a _for loop_.

In [29]:
import math

P = 10
c = 5
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

n = 2 # two disks
L_n = L/n #thickness of each disk

# calculate elgongation of the first disk (the disk with the larger radius)
for i in range(1,n+1):
    L_i = i*L/n-(L_n*(1/2))
    print(f"L({i}) position: {L_i}")
    r_i = (-c/L)*L_i+2*c
    print(f"r({i}) length: {r_i}")
    A_i = math.pi*r_i**2
    d_i = (P*L/n)/(A_i*E)
    print(f"The delfection of the disk {i} = {d_i}")

L(1) position: 25.0
r(1) length: 8.75
The delfection of the disk 1 = 1.7322987003199492e-05
L(2) position: 75.0
r(2) length: 6.25
The delfection of the disk 2 = 3.395305452627101e-05


We can add all of these deflections together in one variable called ```d_t```. We'll use Python's ```+=``` operator to each disk's deflection each time we go through the *for loop*.

In [30]:
import math

P = 10
c = 5
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

n = 2 # two disks
L_n = L/n #thickness of each disk
d_t = 0

# calculate elgongation of the first disk (the disk with the larger radius)
for i in range(1,n+1):
    L_i = i*L/n-(L_n*(1/2))
    print(f"L({i}) position: {L_i}")
    r_i = (-c/L)*L_i+2*c
    print(f"r({i}) length: {r_i}")
    A_i = math.pi*r_i**2
    d_i = (P*L/n)/(A_i*E)
    print(f"The delfection of the disk {i} = {d_i}")
    d_t += d_i

print(f"The total estimated deflection using {n} disks d_t = {d_t}")


L(1) position: 25.0
r(1) length: 8.75
The delfection of the disk 1 = 1.7322987003199492e-05
L(2) position: 75.0
r(2) length: 6.25
The delfection of the disk 2 = 3.395305452627101e-05
The total estimated deflection using 2 disks d_t = 5.12760415294705e-05


To calculate the percent error we'll use:

$$ \%error = \frac{\delta_{approx}-\delta_{exact}}{\delta_{exact}} \times 100\% $$

In [9]:
pct_error = (d_total-d_exact)/d_exact * 100
print(pct_error)

82.22444774334761


Now let's run it again, but use $n=12$

In [10]:
P = 10
c = 5
A = math.pi*c**2
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

d_exact = delta(P,L,A,E)
print(d_exact)
n = 12
d_total=0
for i in range(n):
    Li = (i+0.5)*(L/n)
    ri = 2*c-c*(Li/L)
    Ai = math.pi*ri**2
    d_part = delta(P,Li,Ai,E)
    d_total += d_part
print(d_total)
pct_error = (d_total-d_exact)/d_exact * 100
print(pct_error)

0.0001061032953945969
0.0003896870374271672
267.2713801941078


Run it one more time, but use $n=60$

In [11]:
P = 10
c = 5
A = math.pi*c**2
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa

d_exact = delta(P,L,A,E)
print(d_exact)
n = 60
d_total=0
for i in range(n):
    Li = (i+0.5)*(L/n)
    ri = 2*c-c*(Li/L)
    Ai = math.pi*ri**2
    d_part = delta(P,Li,Ai,E)
    d_total += d_part
print(d_total)
pct_error = (d_total-d_exact)/d_exact * 100
print(pct_error)

0.0001061032953945969
0.0019532831171970356
1740.9259674101534


In [19]:
import math
P = 10
c = 5
A = math.pi*c**2
L = 100
E = 120*1000 #120 GPa = 120 * 1000 MPa
n = 2
d_total = 0
for i in range(1,n+1):
    Li = i*L/n-(L/n*(1/2))
    print(f"L({i}) position: {Li}")
    ri = (-c/L)*Li+2*c
    print(f"r({i}) length: {ri}")
    Ai = math.pi*ri**2
    di = (P*L/n)/(Ai*E)
    print(di)
    d_total += di
print(d_total)

L(1) position: 25.0
r(1) length: 8.75
1.7322987003199492e-05
L(2) position: 75.0
r(2) length: 6.25
3.395305452627101e-05
5.12760415294705e-05


In [13]:
e_exact

5.305164769729845e-05